Max Pool with indices

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64323 Add embedding shape analysis
* **#64121 Max Pool with indices**
* #63530 Add Maxpool to shape analysis / Opinfo

Add support for aten operators which return multiple outputs

Differential Revision: [D30738142](https://our.internmc.facebook.com/intern/diff/D30738142)

Commit:
Max Pool with indices

[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"




[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs


[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs


[ghstack-poisoned]

ZolotukhinM commented on test/test_ops.py:
```diff
@@ -785,8 +785,14 @@ def get_sample():
                             out = variant(get_sample(), *sample.args, **sample.kwargs)
 
                             # TODO: handle multiple outputs
```
Should this TODO be removed with this change?

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -135,6 +135,25 @@ bool symbolicShapeAnalysisTestModeEnabled() {
+      type->cast<ListType>()->getElementType()->cast<IntType>();
+}
+
+TupleTypePtr mapTensorTupleType(
```
Not sure what this function does from the name, do you mind adding a comment explaining it?

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -449,28 +470,21 @@ struct SymbolicShapeAnalyzer {
     }
-    Node* list_construct = output->node();
+    // If it is not a single list construct or constant, bail,
+    // otherwise we cannot analyze its output and it might be modified
```
Nit: is "otherwise" needed here (from grammar perspective)? Also, should this comment be before the previous check?

ZolotukhinM reviewed (approved):
Cool stuff! LGTM!

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs


[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs


[ghstack-poisoned]

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs


[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738142).

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs

Differential Revision: [D30738142](https://our.internmc.facebook.com/intern/diff/D30738142)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738142).

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs

Differential Revision: [D30738142](https://our.internmc.facebook.com/intern/diff/D30738142)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738142).

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs

Differential Revision: [D30738142](https://our.internmc.facebook.com/intern/diff/D30738142)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738142).

Commit:
Update on "Max Pool with indices"


Add support for aten operators which return multiple outputs

Differential Revision: [D30738142](https://our.internmc.facebook.com/intern/diff/D30738142)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738142).
